task/aws/task.go

```diff
@@ -0,0 +1,336 @@
+	if err := t.DataSources.Credentials.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading SecurityGroup...")
+	if err := t.Resources.SecurityGroup.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading KeyPair...")
+	if err := t.Resources.KeyPair.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading Credentials...")
+	if err := t.DataSources.Credentials.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading LaunchTemplate...")
+	if err := t.Resources.LaunchTemplate.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Reading AutoScalingGroup...")
+	if err := t.Resources.AutoScalingGroup.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Done!")
+	t.Attributes.Addresses = t.Resources.AutoScalingGroup.Attributes.Addresses
+	t.Attributes.Status = t.Resources.AutoScalingGroup.Attributes.Status
+	t.Attributes.Events = t.Resources.AutoScalingGroup.Attributes.Events
+	return nil
+}
+
+func (t *Task) Delete(ctx context.Context) error {
+	log.Println("[INFO] Downloading Directory...")
+	if t.Attributes.Environment.Directory != "" && t.Read(ctx) == nil {
+		if err := t.Pull(ctx, t.Attributes.Environment.Directory); err != nil && err != universal.NotFoundError {
+			return err
+		}
+	}
+	log.Println("[INFO] Deleting AutoScalingGroup...")
+	if err := t.Resources.AutoScalingGroup.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting LaunchTemplate...")
+	if err := t.Resources.LaunchTemplate.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting KeyPair...")
+	if err := t.Resources.KeyPair.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting SecurityGroup...")
+	if err := t.Resources.SecurityGroup.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting Credentials...")
+	if err := t.DataSources.Credentials.Read(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting Bucket...")
+	if err := t.Resources.Bucket.Delete(ctx); err != nil {
+		return err
+	}
+	log.Println("[INFO] Deleting ")
+	return nil
+}
+
+func (t *Task) Logs(ctx context.Context) ([]string, error) {
+	if err := t.Read(ctx); err != nil {
+		return nil, err
+	}
+
+	remote, err := fs.NewFs(ctx, (*t.DataSources.Credentials.Resource)["RCLONE_REMOTE"])
+	if err != nil {
+		return nil, err
+	}
+
+	entries, err := remote.List(ctx, "/log/task")
+	if err != nil {
+		return nil, err
+	}
+
+	var logs []string
+	for _, entry := range entries {
+		object, err := remote.NewObject(ctx, entry.Remote())
+		if err != nil {
+			return nil, err
+		}
+		reader, err := object.Open(ctx)
+		if err != nil {
+			return nil, err
+		}
+		buffer := new(bytes.Buffer)
+		if _, err := io.Copy(buffer, reader); err != nil {
+			return nil, err
+		}
+		logs = append(logs, buffer.String())
+		reader.Close()
+	}
+
+	return logs, nil
+}
+
+func (t *Task) Pull(ctx context.Context, destination string) error {
```

Comment:
☹️ ➕ ➕  I know, I know. That was my biggest concern regarding code quality, but it doesn't have an immediate and elegant solution. Neither [embedding](https://golang.org/doc/effective_go#embedding) nor plain functions seem like an ideal short term solution, because we would have to create a package just for code that is common for cloud machines. Adding this in the “general task” is not possible because of circular references.

Reply:
Even if we die together in dispair, I would suggest to keep it as–is and spend a bit of head–scratching time after the release to find an elegant solution.

Reply:
**Related to**
* https://github.com/iterative/terraform-provider-iterative/pull/237#discussion_r748353566
* https://github.com/iterative/terraform-provider-iterative/pull/237#discussion_r749823918

Reply:
> Even if we die together in dispair, I would suggest to keep it as–is and spend a bit of head–scratching time after the release to find an elegant solution.

Very easy. This methods only exists in the Task abstraction and the business logic just use them from there!
I do not see the issue. Another solution could be use your universal library... there are so many options.
I do not fully understand your concerns.

Reply:
> _Very easy. These methods only exists in the Task abstraction and the business logic just use them from there!_

If you mean to add them to the same source file that contains the `Task` interface (`task/task.go`), you're out of luck: see also circular dependency.

Reply:
Adding it to `universal` doesn't seem especially elegant, but at least is better than pointless repetition, I guess.

Reply:
4d8d521e40e6478edfbfab05610c5f318536867d